docs/proposal-001_add-ons.md

```diff
@@ -0,0 +1,151 @@
+- dependencies (supported Kubernetes versions and corresponding cloud providers, list of references to other add-ons)
+- incompatibilities (supported Kubernetes versions and corresponding cloud providers, list of reference to other add-ons)
+
+In other words, it's ought to be possible to define an add-on that only includes:
+
+- dependencies on other add-ons
+- cloud provider resource definitions
+- Kubernetes configuration objects
+- Kubernetes workload definitions
+
+The implementation must also ensure:
+
+- fully deterministic behaviour for a given version of eksctl, thereby it should have
+    - strict versioning of built-in add-ons by default
+    - any externally-sourced config has to be pinned down or vendored *[TBD]*
+    - there should be an option to use latest version or altenative source URL
+- user should be able:
+    - remove any add-on (including it's dependencies) at any time
+    - customise an add-on in an arbitrary fashion
+    - upgrade an add-on
+    - install add-on at any time
+- an add-on may consists exclusively of
+    - a reference to any Helm chart (or a copy of one)
+    - a reference to a Ksonnet application definition
+
+## Phases of Development
+
+1. discuss and specify add-on definition objects and repository layout
+2. discuss and specify how different config management tools would integrate (Helm, K/Jsonnet, kubegen, kustomize)
+3. implement ability to
+    - install a built-in add-on (must ensure upgrades are catered for)
+    - install an external add-on
+    - reconfigure customer an installed add-on
+    - upgrade any installed add-on
+
+## Design Recommendations
+
+- *[TBD]*
+
+- built-in add-ons:
+    - only one version of each add-on is expected to be built-in, alternative versions have to be sourced externally
+    - only built-in add-ons may include Go code to create AWS resources (this may change in the future, yet significantly simplifies initial design and ensures security)
+
+## API
+
+### Add-on instance examples
+
+Install a buil-in `flux` addon:
+
+```YAML
+apiVersion: eksctl.io/v1alpha1
+kind: AddonInstance
+metadata:
+    name: flux
+params:
+    newCodeCommitRepo: true
+    witHelmOperator: true
+```
+
+Install a buil-in `helm` addon:
+
+```YAML
+apiVersion: eksctl.io/v1alpha1
+kind: AddonInstance
+metadata:
+    name: helm
+```
+
+
+Install a external example addon:
+
+```YAML
+apiVersion: eksctl.io/v1alpha1
+kind: AddonInstance
+metadata:
+    name: kustomize-example
+spec:
+    source:
+        repo: https://github.com/weaveworks/eksctl-addons
+        branch: $HEAD
+        path: examples/kustomize
+    kustomization:
+        # bases not included, it's the add-on source
+        patchesStrategicMerge:
+        - patch.yaml
+        namePrefix: example-
+```
+
+### Add-on module examples
+
+```YAML
+apiVersion: eksctl.io/v1alpha1
+kind: AddonModule
+metadata:
+    name: flux
+spec:
+    source:
+        repo: https://github.com/weaveworks/eksctl
+        branch: $HEAD
```

Comment:
Would tags also be supported?

Reply:
Yes, all the usual things would make sense. We might just rely on https://github.com/hashicorp/go-getter actually, it seems to have all the features we'd want.